### PR TITLE
net: add rcvbuf socket option

### PR DESCRIPTION
**Pull request by mohankumarkumar**
Introduce set/get SO_RCVBUF option using the setsockopt
function. In addition, use the rcvbuf value to set the
tcp recv window.

Signed-off-by: Mohan Kumar Kumar <mohankm@fb.com>

### CODE REVIEW

File: subsys/net/ip/net_context.c [hunk 1]
```diff
@@ -2213,6 +2228,22 @@ static int set_context_sndtimeo(struct net_context *context,
 #endif
 }
 
+static int set_context_rcvbuf(struct net_context *context,
+				const void *value, size_t len)
+{
+#if defined(CONFIG_NET_CONTEXT_RCVBUF)
+	if (len != sizeof(uint16_t)) {
```
Comment (rlubos) [inline subsys/net/ip/net_context.c:2235]: I think the expected option type should rather be `int` to be compatible with Linux. We can check here if the value fits within expected range and return error if not.

File: subsys/net/ip/net_context.c [hunk 2]
```diff
@@ -1245,6 +1245,21 @@ static int get_context_sndtimeo(struct net_context *context,
 #endif
 }
 
+static int get_context_rcvbuf(struct net_context *context,
+			      void *value, size_t *len)
+{
+#if defined(CONFIG_NET_CONTEXT_RCVBUF)
+	*((uint16_t *)value) = context->options.rcvbuf;
```
Comment (rlubos) [inline subsys/net/ip/net_context.c:1252]: Ditto.

File: subsys/net/ip/tcp.c [hunk 3]
```diff
@@ -1245,6 +1247,14 @@ static struct tcp *tcp_conn_alloc(void)
 	conn->state = TCP_LISTEN;
 	conn->recv_win = tcp_window;
 
+	/* Set the recv_win with the rcvbuf configured for the socket.
+	 */
```
Comment (rlubos) [inline subsys/net/ip/tcp.c:1251]: Nit, you could just close the comment at the same line, no need to split.

File: subsys/net/lib/sockets/sockets.c [hunk 4]
```diff
@@ -1748,9 +1748,22 @@ int zsock_getsockopt_ctx(struct net_context *ctx, int level, int optname,
...
 		}
-		}
-
 		break;
+
+		case SO_RCVBUF:
+			if (IS_ENABLED(CONFIG_NET_CONTEXT_RCVBUF)) {
+				ret = net_context_get_option(ctx,
+						NET_OPT_RCVBUF,
+						optval, optlen);
```
Comment (rlubos) [inline subsys/net/lib/sockets/sockets.c:1757]: nit: Please align to the opening bracket.

File: subsys/net/lib/sockets/sockets.c [hunk 5]
```diff
@@ -1802,6 +1815,21 @@ int zsock_setsockopt_ctx(struct net_context *ctx, int level, int optname,
 	switch (level) {
 	case SOL_SOCKET:
 		switch (optname) {
+		case SO_RCVBUF:
+			if (IS_ENABLED(CONFIG_NET_CONTEXT_RCVBUF)) {
+				ret = net_context_set_option(ctx,
+						NET_OPT_RCVBUF,
+						optval, optlen);
```
Comment (rlubos) [inline subsys/net/lib/sockets/sockets.c:1822]: nit: Please align to the opening bracket.

File: tests/net/socket/tcp/src/main.c [hunk 6]
```diff
@@ -672,6 +672,39 @@ void test_so_protocol(void)
...
+	rv = setsockopt(sock2, SOL_SOCKET, SO_RCVBUF, &optval, sizeof(optval));
+	zassert_equal(rv, 0, "setsockopt failed (%d)", rv);
+	rv = getsockopt(sock2, SOL_SOCKET, SO_RCVBUF, &retval, &optlen);
+	zassert_equal(rv, 0, "getsockopt failed (%d)", rv);
+	zassert_equal(retval, optval, "getsockopt got invalid rcvbuf");
+	zassert_equal(optlen, sizeof(optval), "getsockopt got invalid size");
+
+	test_close(sock1);
+	test_close(sock2);
+	k_sleep(TCP_TEARDOWN_TIMEOUT);
```
Comment (rlubos) [inline tests/net/socket/tcp/src/main.c:705]: I don't think we need this delay here, we don't even initiate the connection.

**Review by rlubos — COMMENTED** [review]
Looks pretty good, thanks for the contribution! Just a few comments.

File: subsys/net/ip/tcp.c [hunk 7]
```diff
@@ -1245,6 +1247,14 @@ static struct tcp *tcp_conn_alloc(void)
...
+	if (net_context_get_option(context, NET_OPT_RCVBUF, &recv_window, &len) == 0) {
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:1252]: We could add here check whether the rcvbuf option is enabled, so
```
if (IS_ENABLED(CONFIG_NET_CONTEXT_RCVBUF) && net_context_get_option(context, NET_OPT_RCVBUF, &recv_window, &len) == 0) {
```

File: tests/net/socket/tcp/prj.conf [hunk 8]
```diff
@@ -35,3 +35,4 @@ CONFIG_ZTEST=y
 CONFIG_ZTEST_STACK_SIZE=2048
```
Comment (jukkar) [inline tests/net/socket/tcp/prj.conf:35]: Could you split the tests to a separate commit.

**Comment by mohankumarkumar** [comment]
Thanks for the comments! Addressed all the comments in the latest commits.

**Review by jukkar — APPROVED** [review]

**Review by rlubos — APPROVED** [review]
Thanks, LGTM
